File: templates/cpp/protocols/heartbeat_protocol.cpp
#include "heartbeat_protocol.hpp"

#include <condition_variable>
#include <cstring>

namespace mavlink {

HeartbeatMonitor::HeartbeatMonitor(
  MavlinkSession* session,
  std::chrono::milliseconds timeout,
  std::optional<std::set<MavlinkNode>> watch,
  std::optional<uint8_t> watch_system_id
)
    : session_(session),
      timeout_(timeout),
      watch_(std::move(watch)),
      watch_system_id_(watch_system_id) {}

HeartbeatMonitor::~HeartbeatMonitor() { stop(); }

void HeartbeatMonitor::start() {
  std::lock_guard<std::mutex> lock(mutex_);
  if (running_) {
    return;
  }
  running_ = true;
  frame_listener_id_ = session_->add_frame_listener([this](const frame_t& frame) { on_frame(frame); });

  watchdog_thread_ = std::make_unique<std::thread>([this]() {
    while (running_) {
      std::this_thread::sleep_for(timeout_ / 3);
      if (!running_) {
        break;
      }
      check_timeouts();
    }
  });
}

void HeartbeatMonitor::stop() {
  {
    std::lock_guard<std::mutex> lock(mutex_);
    if (!running_) {
      return;
    }
    running_ = false;
  }

  if (watchdog_thread_ && watchdog_thread_->joinable()) {
    watchdog_thread_->join();
  }
  watchdog_thread_.reset();
  session_->remove_frame_listener(frame_listener_id_);
}

void HeartbeatMonitor::on_heartbeat(heartbeat_callback_t callback) {
  std::lock_guard<std::mutex> lock(mutex_);
  heartbeat_callbacks_.push_back(std::move(callback));
}

void HeartbeatMonitor::on_connected(node_callback_t callback) {
  std::lock_guard<std::mutex> lock(mutex_);
  connected_callbacks_.push_back(std::move(callback));
}

void HeartbeatMonitor::on_disconnected(node_callback_t callback) {
  std::lock_guard<std::mutex> lock(mutex_);
  disconnected_callbacks_.push_back(std::move(callback));
}

std::optional<TrackedHeartbeat> HeartbeatMonitor::state_for(const MavlinkNode& node) const {
  std::lock_guard<std::mutex> lock(mutex_);
  const auto it = states_.find(node);
  if (it == states_.end()) {
    return std::nullopt;
  }
  return it->second;
}

std::optional<TrackedHeartbeat> HeartbeatMonitor::state_for_ids(
  uint8_t system_id,
  uint8_t component_id
) const {
  return state_for(MavlinkNode{system_id, component_id});
}

bool HeartbeatMonitor::is_online(const MavlinkNode& node) const {
  std::lock_guard<std::mutex> lock(mutex_);
  const auto it = online_.find(node);
  return it != online_.end() && it->second;
}

bool HeartbeatMonitor::is_online_ids(uint8_t system_id, uint8_t component_id) const {
  return is_online(MavlinkNode{system_id, component_id});
}

std::vector<MavlinkNode> HeartbeatMonitor::online_nodes() const {
  std::lock_guard<std::mutex> lock(mutex_);
  std::vector<MavlinkNode> nodes;
  for (const auto& entry : online_) {
    if (entry.second) {
      nodes.push_back(entry.first);
    }
  }
  return nodes;
}

MavlinkNode HeartbeatMonitor::wait_for_vehicle(
  const std::set<uint8_t>* exclude_system_ids,
  std::chrono::milliseconds timeout,
  MavlinkCancellationToken* cancel
) {
  if (cancel != nullptr) {
    cancel->throw_if_cancelled();
  }

  for (const auto& node : online_nodes()) {
    if (exclude_system_ids == nullptr || exclude_system_ids->count(node.system_id) == 0) {
      return node;
    }
  }

  std::mutex wait_mutex;
  std::condition_variable cv;
  bool done = false;
  MavlinkNode result{0, 0};

  auto connected_cb = [&](const MavlinkNode& node) {
    if (exclude_system_ids != nullptr && exclude_system_ids->count(node.system_id) != 0) {
      return;
    }
    std::lock_guard<std::mutex> lock(wait_mutex);
    if (!done) {
      done = true;
      result = node;
      cv.notify_all();
    }
  };

  on_connected(connected_cb);

  std::unique_lock<std::mutex> lock(wait_mutex);
  if (cancel != nullptr) {
    cancel->on_cancel([&]() {
      std::lock_guard<std::mutex> inner(wait_mutex);
      done = true;
      cv.notify_all();
    });
  }

  if (!cv.wait_for(lock, timeout, [&]() { return done; })) {
    throw MavlinkTimeoutException("Timed out waiting for vehicle heartbeat", timeout);
  }

  if (cancel != nullptr && cancel->is_cancelled()) {
    throw MavlinkCancelledException();
  }

  return result;
}

void HeartbeatMonitor::on_frame(const frame_t& frame) {
  if (frame.message_id != heartbeat_MSG_ID) {
    return;
  }

  heartbeat_t heartbeat_msg{};
  heartbeat_parse(frame.payload, heartbeat_msg);

  const MavlinkNode node{frame.system_id, frame.component_id};
  if (!should_watch(node)) {
    return;
  }

  std::vector<heartbeat_callback_t> heartbeat_cbs;
  std::vector<node_callback_t> connected_cbs;
  TrackedHeartbeat tracked{};

  {
    std::lock_guard<std::mutex> lock(mutex_);
    const bool was_online = online_.count(node) != 0 && online_[node];
    tracked = TrackedHeartbeat{
      node,
      heartbeat_msg,
      std::chrono::steady_clock::now(),
      true,
    };
    states_[node] = tracked;
    online_[node] = true;
    heartbeat_cbs = heartbeat_callbacks_;

    if (!was_online) {
      connected_cbs = connected_callbacks_;
    }
  }

  for (const auto& cb : heartbeat_cbs) {
    if (cb) {
      cb(tracked);
    }
  }
  for (const auto& cb : connected_cbs) {
    if (cb) {
      cb(node);
    }
  }
}

void HeartbeatMonitor::check_timeouts() {
  const auto now = std::chrono::steady_clock::now();
  std::vector<node_callback_t> disconnected_cbs;
  std::vector<heartbeat_callback_t> heartbeat_cbs;
  std::vector<TrackedHeartbeat> offline_updates;

  {
    std::lock_guard<std::mutex> lock(mutex_);
    for (const auto& entry : states_) {
      const auto& node = entry.first;
      const auto& state = entry.second;
      const bool timed_out = now - state.received_at > timeout_;
      const bool was_online = online_.count(node) != 0 && online_[node];

      if (timed_out && was_online) {
        online_[node] = false;
        offline_updates.push_back(TrackedHeartbeat{
          node,
          state.heartbeat_msg,
          state.received_at,
          false,
        });
        disconnected_cbs = disconnected_callbacks_;
        heartbeat_cbs = heartbeat_callbacks_;
      }
    }
  }

  for (const auto& tracked : offline_updates) {
    for (const auto& cb : heartbeat_cbs) {
      if (cb) {
        cb(tracked);
      }
    }
    for (const auto& cb : disconnected_cbs) {
      if (cb) {
        cb(tracked.node);
      }
    }
  }
}

bool HeartbeatMonitor::should_watch(const MavlinkNode& node) const {
  if (watch_.has_value()) {
    return watch_->count(node) != 0;
  }
  if (watch_system_id_.has_value()) {
    return node.system_id == watch_system_id_.value();
  }
  return true;
}

HeartbeatPublisher::HeartbeatPublisher(
  MavlinkSession* session,
  const heartbeat_t& heartbeat_msg,
  std::chrono::milliseconds interval
)
    : session_(session), heartbeat_msg_(heartbeat_msg), interval_(interval) {}

HeartbeatPublisher::~HeartbeatPublisher() { stop(); }

const heartbeat_t& HeartbeatPublisher::heartbeat_msg() const { return heartbeat_msg_; }

void HeartbeatPublisher::update_heartbeat(const heartbeat_t& heartbeat_msg) {
  heartbeat_msg_ = heartbeat_msg;
}

void HeartbeatPublisher::mutate_heartbeat(const std::function<void(heartbeat_t&)>& transform) {
  transform(heartbeat_msg_);
}

void HeartbeatPublisher::start() {
  if (running_.exchange(true)) {
    return;
  }
  send_once();
  thread_ = std::make_unique<std::thread>([this]() { publish_loop(); });
}

void HeartbeatPublisher::stop() {
  if (!running_.exchange(false)) {
    return;
  }
  if (thread_ && thread_->joinable()) {
    thread_->join();
  }
  thread_.reset();
}

void HeartbeatPublisher::send_once() {
  uint8_t payload[heartbeat_ENCODED_LENGTH];
  heartbeat_serialize(heartbeat_msg_, payload);
  session_->send_frame(
    heartbeat_MSG_ID,
    heartbeat_CRC_EXTRA,
    payload,
    heartbeat_ENCODED_LENGTH
  );
}

void HeartbeatPublisher::publish_loop() {
  while (running_) {
    std::this_thread::sleep_for(interval_);
    if (!running_) {
      break;
    }
    send_once();
  }
}

heartbeat_t HeartbeatTemplates::gcs(int mavlink_version) {
  heartbeat_t hb{};
  hb.custom_mode = 0;
  hb.type = MAV_TYPE_GCS;
  hb.autopilot = MAV_AUTOPILOT_INVALID;
  hb.base_mode = 0;
  hb.system_status = MAV_STATE_ACTIVE;
  hb.mavlink_version = static_cast<uint8_t>(mavlink_version);
  return hb;
}

heartbeat_t HeartbeatTemplates::autopilot(
  int mavlink_version,
  MAV_TYPE type,
  MAV_AUTOPILOT autopilot,
  MAV_STATE system_status,
  uint32_t custom_mode,
  uint8_t base_mode
) {
  heartbeat_t hb{};
  hb.custom_mode = custom_mode;
  hb.type = type;
  hb.autopilot = autopilot;
  hb.base_mode = base_mode;
  hb.system_status = system_status;
  hb.mavlink_version = static_cast<uint8_t>(mavlink_version);
  return hb;
}

heartbeat_t HeartbeatTemplates::onboard_api(int mavlink_version) {
  heartbeat_t hb{};
  hb.custom_mode = 0;
  hb.type = MAV_TYPE_ONBOARD_CONTROLLER;
  hb.autopilot = MAV_AUTOPILOT_INVALID;
  hb.base_mode = 0;
  hb.system_status = MAV_STATE_ACTIVE;
  hb.mavlink_version = static_cast<uint8_t>(mavlink_version);
  return hb;
}

}  // namespace mavlink

File: templates/rust/protocols/heartbeat_protocol.rs
//! HEARTBEAT monitoring and publishing.

use std::collections::{HashMap, HashSet};
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

use futures::StreamExt;
use tokio::sync::{broadcast, oneshot};
use tokio::task::JoinHandle;
use tokio::time::{self, MissedTickBehavior};

use crate::{Heartbeat, MavAutopilot, MavState, MavType};
use super::mavlink_cancellation::MavlinkCancellationToken;
use super::mavlink_session::{MavlinkSession, MavlinkTimeoutError, SessionWaitError};

/// MAVLink node identity (system + component).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct MavlinkNode {
    pub system_id: u8,
    pub component_id: u8,
}

impl MavlinkNode {
    pub fn new(system_id: u8, component_id: u8) -> Self {
        Self {
            system_id,
            component_id,
        }
    }
}

impl std::fmt::Display for MavlinkNode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "MavlinkNode({}:{})", self.system_id, self.component_id)
    }
}

/// Last known heartbeat state for a remote node.
#[derive(Debug, Clone)]
pub struct TrackedHeartbeat {
    pub node: MavlinkNode,
    pub heartbeat: Heartbeat,
    pub received_at: Instant,
    pub online: bool,
}

impl TrackedHeartbeat {
    pub fn age(&self) -> Duration {
        self.received_at.elapsed()
    }
}

struct MonitorInner {
    session: Arc<MavlinkSession>,
    timeout: Duration,
    watch: Option<HashSet<MavlinkNode>>,
    watch_system_id: Option<u8>,
    states: Mutex<HashMap<MavlinkNode, TrackedHeartbeat>>,
    online: Mutex<HashMap<MavlinkNode, bool>>,
    heartbeat_tx: broadcast::Sender<TrackedHeartbeat>,
    connected_tx: broadcast::Sender<MavlinkNode>,
    disconnected_tx: broadcast::Sender<MavlinkNode>,
    running: Mutex<bool>,
    reader: Mutex<Option<JoinHandle<()>>>,
    watchdog: Mutex<Option<JoinHandle<()>>>,
}

/// Tracks remote HEARTBEAT messages and reports connect / disconnect events.
pub struct HeartbeatMonitor {
    inner: Arc<MonitorInner>,
}

impl HeartbeatMonitor {
    pub fn new(
        session: Arc<MavlinkSession>,
        timeout: Duration,
        watch: Option<HashSet<MavlinkNode>>,
        watch_system_id: Option<u8>,
    ) -> Self {
        let (heartbeat_tx, _) = broadcast::channel(64);
        let (connected_tx, _) = broadcast::channel(64);
        let (disconnected_tx, _) = broadcast::channel(64);
        Self {
            inner: Arc::new(MonitorInner {
                session,
                timeout,
                watch,
                watch_system_id,
                states: Mutex::new(HashMap::new()),
                online: Mutex::new(HashMap::new()),
                heartbeat_tx,
                connected_tx,
                disconnected_tx,
                running: Mutex::new(false),
                reader: Mutex::new(None),
                watchdog: Mutex::new(None),
            }),
        }
    }

    pub fn on_heartbeat(&self) -> broadcast::Receiver<TrackedHeartbeat> {
        self.inner.heartbeat_tx.subscribe()
    }

    pub fn on_connected(&self) -> broadcast::Receiver<MavlinkNode> {
        self.inner.connected_tx.subscribe()
    }

    pub fn on_disconnected(&self) -> broadcast::Receiver<MavlinkNode> {
        self.inner.disconnected_tx.subscribe()
    }

    pub fn start(&self) {
        let mut running = self.inner.running.lock().unwrap();
        if *running {
            return;
        }
        *running = true;

        let inner = Arc::clone(&self.inner);
        let reader = tokio::spawn(async move {
            let mut frames = inner.session.frames();
            while let Some(frame) = frames.next().await {
                if !*inner.running.lock().unwrap() {
                    break;
                }
                let Some(heartbeat) = (frame.message.as_ref() as &dyn std::any::Any)
                    .downcast_ref::<Heartbeat>()
                    .cloned()
                else {
                    continue;
                };
                let node = MavlinkNode::new(frame.system_id, frame.component_id);
                if !should_watch(&inner, node) {
                    continue;
                }

                let was_online = *inner.online.lock().unwrap().get(&node).unwrap_or(&false);
                let tracked = TrackedHeartbeat {
                    node,
                    heartbeat,
                    received_at: Instant::now(),
                    online: true,
                };
                inner.states.lock().unwrap().insert(node, tracked.clone());
                inner.online.lock().unwrap().insert(node, true);
                let _ = inner.heartbeat_tx.send(tracked);
                if !was_online {
                    let _ = inner.connected_tx.send(node);
                }
            }
        });
        *self.inner.reader.lock().unwrap() = Some(reader);

        let inner = Arc::clone(&self.inner);
        let period = self.inner.timeout.div_f32(3.0).max(Duration::from_millis(100));
        let watchdog = tokio::spawn(async move {
            let mut ticker = time::interval(period);
            ticker.set_missed_tick_behavior(MissedTickBehavior::Skip);
            loop {
                ticker.tick().await;
                if !*inner.running.lock().unwrap() {
                    break;
                }
                check_timeouts(&inner);
            }
        });
        *self.inner.watchdog.lock().unwrap() = Some(watchdog);
    }

    pub async fn stop(&self) {
        let mut running = self.inner.running.lock().unwrap();
        if !*running {
            return;
        }
        *running = false;
        if let Some(handle) = self.inner.reader.lock().unwrap().take() {
            handle.abort();
        }
        if let Some(handle) = self.inner.watchdog.lock().unwrap().take() {
            handle.abort();
        }
    }

    pub fn state_for(&self, node: MavlinkNode) -> Option<TrackedHeartbeat> {
        self.inner.states.lock().unwrap().get(&node).cloned()
    }

    pub fn state_for_ids(&self, system_id: u8, component_id: u8) -> Option<TrackedHeartbeat> {
        self.state_for(MavlinkNode::new(system_id, component_id))
    }

    pub fn is_online(&self, node: MavlinkNode) -> bool {
        *self.inner.online.lock().unwrap().get(&node).unwrap_or(&false)
    }

    pub fn is_online_ids(&self, system_id: u8, component_id: u8) -> bool {
        self.is_online(MavlinkNode::new(system_id, component_id))
    }

    pub fn online_nodes(&self) -> Vec<MavlinkNode> {
        self.inner
            .online
            .lock()
            .unwrap()
            .iter()
            .filter_map(|(node, online)| online.then_some(*node))
            .collect()
    }

    pub async fn wait_for_vehicle(
        &self,
        exclude_system_ids: Option<HashSet<u8>>,
        timeout: Duration,
        cancel: Option<&MavlinkCancellationToken>,
    ) -> Result<MavlinkNode, SessionWaitError> {
        if let Some(token) = cancel {
            token.throw_if_cancelled().map_err(SessionWaitError::Cancelled)?;
        }

        for node in self.online_nodes() {
            if exclude_system_ids
                .as_ref()
                .is_none_or(|ids| !ids.contains(&node.system_id))
            {
                return Ok(node);
            }
        }

        let (tx, rx) = oneshot::channel();
        let mut connected = self.on_connected();
        let waiter = tokio::spawn(async move {
            while let Ok(node) = connected.recv().await {
                if exclude_system_ids
                    .as_ref()
                    .is_some_and(|ids| ids.contains(&node.system_id))
                {
                    continue;
                }
                let _ = tx.send(node);
                break;
            }
        });

        let result = time::timeout(timeout, rx).await;
        waiter.abort();

        match result {
            Ok(Ok(node)) => Ok(node),
            Ok(Err(_)) => Err(SessionWaitError::Closed),
            Err(_) => Err(SessionWaitError::Timeout(MavlinkTimeoutError::new(
                "Timed out waiting for vehicle heartbeat",
                timeout,
            ))),
        }
    }
}

fn should_watch(inner: &MonitorInner, node: MavlinkNode) -> bool {
    if let Some(watch) = &inner.watch {
        return watch.contains(&node);
    }
    if let Some(system_id) = inner.watch_system_id {
        return node.system_id == system_id;
    }
    true
}

fn check_timeouts(inner: &MonitorInner) {
    let now = Instant::now();
    let nodes: Vec<_> = inner.states.lock().unwrap().keys().copied().collect();
    for node in nodes {
        let Some(state) = inner.states.lock().unwrap().get(&node).cloned() else {
            continue;
        };
        let timed_out = now.duration_since(state.received_at) > inner.timeout;
        let was_online = *inner.online.lock().unwrap().get(&node).unwrap_or(&false);
        if timed_out && was_online {
            inner.online.lock().unwrap().insert(node, false);
            let _ = inner.disconnected_tx.send(node);
            let _ = inner.heartbeat_tx.send(TrackedHeartbeat {
                node,
                heartbeat: state.heartbeat,
                received_at: state.received_at,
                online: false,
            });
        }
    }
}

/// Periodically sends HEARTBEAT on a [`MavlinkSession`].
pub struct HeartbeatPublisher {
    session: Arc<MavlinkSession>,
    interval: Duration,
    heartbeat: Mutex<Heartbeat>,
    timer: Mutex<Option<JoinHandle<()>>>,
    running: Mutex<bool>,
}

impl HeartbeatPublisher {
    pub fn new(session: Arc<MavlinkSession>, heartbeat: Heartbeat, interval: Duration) -> Self {
        Self {
            session,
            interval,
            heartbeat: Mutex::new(heartbeat),
            timer: Mutex::new(None),
            running: Mutex::new(false),
        }
    }

    pub fn heartbeat(&self) -> Heartbeat {
        self.heartbeat.lock().unwrap().clone()
    }

    pub fn update_heartbeat(&self, heartbeat: Heartbeat) {
        *self.heartbeat.lock().unwrap() = heartbeat;
    }

    pub fn mutate_heartbeat(&self, transform: impl FnOnce(Heartbeat) -> Heartbeat) {
        let mut heartbeat = self.heartbeat.lock().unwrap();
        *heartbeat = transform(heartbeat.clone());
    }

    pub fn start(&self) {
        let mut running = self.running.lock().unwrap();
        if *running {
            return;
        }
        *running = true;
        let session = Arc::clone(&self.session);
        let interval = self.interval;
        let heartbeat = Arc::new(Mutex::new(self.heartbeat.lock().unwrap().clone()));
        let running_flag = Arc::new(Mutex::new(true));
        let timer_running = Arc::clone(&running_flag);
        let timer = tokio::spawn(async move {
            let _ = send_once(&session, &heartbeat).await;
            let mut ticker = time::interval(interval);
            ticker.set_missed_tick_behavior(MissedTickBehavior::Skip);
            loop {
                ticker.tick().await;
                if !*timer_running.lock().unwrap() {
                    break;
                }
                let _ = send_once(&session, &heartbeat).await;
            }
        });
        *self.timer.lock().unwrap() = Some(timer);
    }

    pub fn stop(&self) {
        *self.running.lock().unwrap() = false;
        if let Some(handle) = self.timer.lock().unwrap().take() {
            handle.abort();
        }
    }

    pub async fn send_once(&self) -> Result<(), SessionWaitError> {
        send_once(&self.session, &self.heartbeat).await
    }
}

async fn send_once(
    session: &MavlinkSession,
    heartbeat: &Mutex<Heartbeat>,
) -> Result<(), SessionWaitError> {
    let heartbeat = heartbeat.lock().unwrap().clone();
    session.send(Box::new(heartbeat)).await
}

/// Convenience factories for common HEARTBEAT payloads.
pub struct HeartbeatTemplates;

impl HeartbeatTemplates {
    pub fn gcs(mavlink_version: u8) -> Heartbeat {
        Heartbeat {
            custom_mode: 0,
            r#type: MavType::MAV_TYPE_GCS,
            autopilot: MavAutopilot::MAV_AUTOPILOT_INVALID,
            base_mode: 0,
            system_status: MavState::MAV_STATE_ACTIVE,
            mavlink_version,
        }
    }

    pub fn autopilot(
        mavlink_version: u8,
        vehicle_type: MavType,
        autopilot: MavAutopilot,
        system_status: MavState,
        custom_mode: u32,
        base_mode: u8,
    ) -> Heartbeat {
        Heartbeat {
            custom_mode,
            r#type: vehicle_type,
            autopilot,
            base_mode,
            system_status,
            mavlink_version,
        }
    }

    pub fn autopilot_default(mavlink_version: u8) -> Heartbeat {
        Self::autopilot(
            mavlink_version,
            MavType::MAV_TYPE_QUADROTOR,
            MavAutopilot::MAV_AUTOPILOT_PX4,
            MavState::MAV_STATE_ACTIVE,
            0,
            0,
        )
    }

    pub fn onboard_api(mavlink_version: u8) -> Heartbeat {
        Heartbeat {
            custom_mode: 0,
            r#type: MavType::MAV_TYPE_ONBOARD_CONTROLLER,
            autopilot: MavAutopilot::MAV_AUTOPILOT_INVALID,
            base_mode: 0,
            system_status: MavState::MAV_STATE_ACTIVE,
            mavlink_version,
        }
    }
}
